sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java

```diff
@@ -906,19 +946,123 @@ public void setValueDeserializer(String valueDeserializer) {
+                .withOffsetConsumerConfigOverrides(getOffsetConsumerConfig())
+                .withConsumerFactoryFn(getConsumerFactoryFn())
+                .withKeyDeserializerProvider(getKeyDeserializerProvider())
+                .withValueDeserializerProvider(getValueDeserializerProvider())
+                .withExtractOutputTimestampFn(timestampFn);
+        if (isCommitOffsetsInFinalizeEnabled()) {
+          readTransform = readTransform.commitOffsets();
+        }
+
+        return input
+            .getPipeline()
+            .apply(Impulse.create())
+            .apply(
+                ParDo.of(
+                    new GenerateKafkaSourceDescription(
+                        readTransform.getKafkaSourceDescriptionSchema())))
+            .setCoder(RowCoder.of(readTransform.getKafkaSourceDescriptionSchema()))
+            .apply(readTransform)
+            .setCoder(KafkaRecordCoder.of(keyCoder, valueCoder));
+      }
+    }
+
+    /**
+     * A DoFn which generates {@link Row} with {@link KafkaSourceDescriptionSchemas#getSchema()}
+     * based on the configuration of {@link Read}.
+     */
+    @VisibleForTesting
+    class GenerateKafkaSourceDescription extends DoFn<byte[], Row> {
+      GenerateKafkaSourceDescription(Schema schema) {
+        this.kafkaSourceDescriptionSchema = schema;
+      }
+
+      private final Schema kafkaSourceDescriptionSchema;
+
+      private final Map<String, Object> consumerConfig = Read.this.getConsumerConfig();
 
-      PTransform<PBegin, PCollection<KafkaRecord<K, V>>> transform = unbounded;
+      private final SerializableFunction<Map<String, Object>, Consumer<byte[], byte[]>>
+          consumerFactoryFn = Read.this.getConsumerFactoryFn();
 
-      if (getMaxNumRecords() < Long.MAX_VALUE || getMaxReadTime() != null) {
-        transform =
-            unbounded.withMaxReadTime(getMaxReadTime()).withMaxNumRecords(getMaxNumRecords());
+      private final List<String> topics = Read.this.getTopics();
+
+      private final List<TopicPartition> topicPartitions = Read.this.getTopicPartitions();
+
+      private final Instant startReadTime = Read.this.getStartReadTime();
+
+      @VisibleForTesting
+      Map<String, Object> getConsumerConfig() {
+        return consumerConfig;
       }
 
-      return input.getPipeline().apply(transform);
+      @VisibleForTesting
+      List<String> getTopics() {
+        return topics;
+      }
+
+      @ProcessElement
+      public void processElement(OutputReceiver<Row> receiver) {
+        List<TopicPartition> partitions = new ArrayList<>(topicPartitions);
+        if (partitions.isEmpty()) {
+          try (Consumer<?, ?> consumer = consumerFactoryFn.apply(consumerConfig)) {
+            for (String topic : topics) {
+              for (PartitionInfo p : consumer.partitionsFor(topic)) {
+                partitions.add(new TopicPartition(p.topic(), p.partition()));
+              }
+            }
+          }
+        }
+        partitions.stream()
```

Comment:
nit: this is much harder to read then a for loop. Also the performance of stream() is poor relative to the for loop as well.